R/R/decorators-aws.R

```diff
@@ -0,0 +1,125 @@
+#' @param iam_role Character. IAM role that AWS Batch can use to access Amazon
+#'   S3. Defaults to the one determined by the environment variable
+#'   METAFLOW_ECS_S3_ACCESS_IAM_ROLE
+#' @param execution_role Character. IAM role that AWS Batch can use to trigger
+#'   AWS Fargate tasks. Defaults to the one determined by the environment
+#'   variable METAFLOW_ECS_FARGATE_EXECUTION_ROLE. See the
+#'   \href{https://docs.aws.amazon.com/batch/latest/userguide/execution-IAM-role.html}{AWS
+#'    Documentation} for more information.
+#' @param shared_memory Integer. The value for the size (in MiB) of the
+#'   `/dev/shm` volume for this step. This parameter maps to the `--shm-size`
+#'   option to `docker run`.
+#' @param max_swap Integer. The total amount of swap memory (in MiB) a container
+#'   can use for this step. This parameter is translated to the `--memory-swap`
+#'   option to docker run where the value is the sum of the container memory
+#'   plus the `max_swap` value.
+#' @param swappiness This allows you to tune memory swappiness behavior for this
+#'   step. A swappiness value of `0` causes swapping not to happen unless
+#'   absolutely necessary. A swappiness value of `100` causes pages to be
+#'   swapped very aggressively. Accepted values are whole numbers between `0`
+#'   and `100`.
+#'   
+#' @inherit decorator return
+#'
+#' @export
+#' 
+#' @examples \dontrun{
+#' # This example will generate a large random matrix which takes up roughly 
+#' # 48GB of memory, and sums the entries. The `batch` decorator forces this
+#' # step to run in an environment with 60000MB of memory.
+#' 
+#' start <- function(self) {
+#'   big_matrix <- matrix(rexp(80000*80000), 80000)
+#'   self$sum <- sum(big_matrix)
+#' }
+#' 
+#' end <- function(self) {
+#'   message(
+#'     "sum is: ", self$sum
+#'   )
+#' }
+#' 
+#' metaflow("BigSumFlowR") %>%
+#'   step(
+#'     batch(memory=60000, cpu=1),
+#'     step = "start",
+#'     r_function = start,
+#'     next_step = "end"
+#'   ) %>%
+#'   step(
+#'     step = "end",
+#'     r_function = end
+#'   ) %>%
+#'   run()
+#' }
+batch <- function(
+  cpu = 1L,
+  gpu = 0L,
+  memory = 4096L,
+  image = NULL,
+  queue = NULL,
+  iam_role = NULL,
+  execution_role = NULL,
+  shared_memory = NULL,
+  max_swap = NULL,
+  swappiness = NULL
+) {
+  queue = queue %||% pkg.env$mf$metaflow_config$BATCH_JOB_QUEUE
+  iam_role = iam_role %||% pkg.env$mf$metaflow_config$ECS_S3_ACCESS_IAM_ROLE
+  execution_role = execution_role %||% pkg.env$mf$metaflow_config$ECS_FARGATE_EXECUTION_ROLE
+  
+  decorator(
+    "batch",
+    cpu = cpu,
+    gpu = gpu,
+    memory = memory,
+    image = image,
+    queue = queue,
+    iam_role = iam_role,
+    execution_role = execution_role,
+    shared_memory = shared_memory,
+    max_swap = max_swap,
+    swappiness = swappiness
+  )
+}
+
+#' @rdname batch
+#' @export
+resources <- function(
+  cpu = 1L,
+  gpu = 0L,
+  memory = 4096L,
+  shared_memory = NULL
```

Comment:
the argument list should be the same as the list for batch?

Reply:
My reading of [the (Python) source code](https://github.com/Netflix/metaflow/blob/4b82c586b218efe226be1dd88a1dc253b5c24e63/metaflow/plugins/aws/batch/batch_decorator.py#L33) is that `@resources` doesn't take the same arguments. To verify, I tried a Python flow with a `@resources(max_swap = 100)` and I got:

```python
metaflow.exception.InvalidDecoratorAttribute: Decorator 'resources' does not support the attribute 'max_swap'. These attributes are supported: cpu, gpu, memory, shared_memory.
```